Clean up datetime operators

martint opened the pull request.

martint committed:
Clarify variable names in DateTimeOperators

martint committed:
Simplify implementation of datetime operations

Implement subtractions in terms of addition with a negative term,
and commutative variants in terms of a canonical implementation.

martint committed:
Clarify parsing of timestamps and time types

Separate between parse (strict) vs convert (which can include a timezone) for timestamp types.
Parsing is used when we know it's a timestamp. convertXXX is used for the purpose of casting
from a datetime string to a timestamp.

martint committed:
Fix timestamp CAST to JSON in new semantics

findepi commented on presto-main/src/main/java/io/prestosql/util/DateTimeUtils.java:
```diff
@@ -143,58 +143,37 @@ public static String printDate(int days)
-     *
-     * @return stack representation of legacy TIMESTAMP or TIMESTAMP WITH TIME ZONE type, depending on input
-     * @deprecated applicable in legacy timestamp semantics only
+     * Parse a string (optionally containing a zone) as a value of TIMESTAMP type using legacy semantics.
```
Specify in the javadoc what happens when

- timestamp is a gap in the provided zone (rejected or shifted)
- value contains a zone specifier -- is it converted to provided zone with "same zone" or "same instant" semantics
  - and what happens if it cannot be

findepi commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/util/KafkaLoader.java:
```diff
@@ -143,13 +143,13 @@ private Object convertValue(Object value, Type type)
             }
             if (TIME.equals(type)) {
-                return ISO8601_FORMATTER.print(parseTimeLiteral(timeZoneKey, (String) value));
+                return ISO8601_FORMATTER.print(parseLegacyTime(timeZoneKey, (String) value));
```
BTW this looks like you're fixing a bug. Previously the code was using `parseTimeLiteral` which `@return stack representation of legacy TIME or TIME WITH TIME ZONE type, depending on input`, yet it was handling the return value as being of TIME type.

As such, this deserves a separate commit and a test.

martint replied:
I'm not really sure what the semantics would be. Let me look into it more closely.

martint replied:
BTW, there's no "convert" semantics for TIME. We either required a timezone (for TIME w/ TZ) or did not allow it (for TIME w/o TZ). We don't currently do the same thing as timestamps, where we allow optional TZ for TIMESTAMP w/ TZ.

martint replied:
I'll fix this in a separate commit (calling the appropriate parseXXX method), but I don't know how to add a test for this.

findepi replied:
I don't know either, but i would start from `io.prestosql.plugin.kafka.TestMinimalFunctionality#createMessages` (but in `TestKafkaIntegrationSmokeTest` class)

findepi commented on presto-main/src/main/java/io/prestosql/sql/planner/LiteralInterpreter.java:
```diff
@@ -151,27 +164,38 @@ protected Object visitGenericLiteral(GenericLiteral node, ConnectorSession sessi
+                }
+                return parseTimeWithoutTimeZone(node.getValue());
+            }
+            else if (type instanceof TimeWithTimeZoneType) {
```
redundant else, below too

findepi commented on presto-main/src/main/java/io/prestosql/sql/planner/LiteralInterpreter.java:
```diff
@@ -151,27 +164,38 @@ protected Object visitGenericLiteral(GenericLiteral node, ConnectorSession sessi
             }
-            catch (RuntimeException e) {
-                throw semanticException(INVALID_LITERAL, node, "'%s' is not a valid timestamp literal", node.getValue());
+            else if (type instanceof TimestampWithTimeZoneType) {
```
redundant else

findepi commented on presto-main/src/main/java/io/prestosql/util/DateTimeUtils.java:
```diff
@@ -217,19 +196,35 @@ public static long parseTimestampWithoutTimeZone(String value)
+        return packDateTimeWithZone(dateTime);
     }
 
+    /**
+     * Parse a string (optionally containing a zone) as a value of TIMESTAMP type.
+     * If the string doesn't specify a zone, it is interpreted in {@code timeZoneKey} zone.
+     *
+     * @return stack representation of legacy TIMESTAMP type
+     * @deprecated applicable in legacy timestamp semantics only
+     */
```
This is a javadoc of something else, not this method.

findepi commented on presto-main/src/main/java/io/prestosql/util/DateTimeUtils.java:
```diff
@@ -217,19 +196,35 @@ public static long parseTimestampWithoutTimeZone(String value)
     {
-        return TIMESTAMP_WITH_OR_WITHOUT_TIME_ZONE_FORMATTER.withChronology(getChronology(timeZoneKey)).parseMillis(value);
+        DateTime dateTime = TIMESTAMP_WITH_OR_WITHOUT_TIME_ZONE_FORMATTER.withChronology(getChronology(timeZoneKey)).withOffsetParsed().parseDateTime(timestampWithTimeZone);
+        return packDateTimeWithZone(dateTime);
```
The code here used to be known as `parseTimestampWithTimeZone(TimeZoneKey timeZoneKey, String timestampWithTimeZone)`, but in the diff it appears as a new code.

martint replied:
Ah, that's because I inadvertently moved methods around. I'll restore the order to make the diff easier to review.

findepi commented on presto-main/src/main/java/io/prestosql/operator/scalar/JsonOperators.java:
```diff
@@ -340,7 +340,12 @@ public static Slice castFromTimestamp(ConnectorSession session, @SqlType(TIMESTA
+                    jsonGenerator.writeString(printTimestampWithoutTimeZone(session.getTimeZoneKey(), value));
+                }
+                else {
+                    jsonGenerator.writeString(printTimestampWithoutTimeZone(value));
```
> Fix legacy/new timestamp cast to json

This should probably be "Fix timestamp CAST to JSON in new semantics" or something, as you are not fixing anything in the legacy semantics (right?)

martint replied:
Yes

findepi reviewed (changes requested):
Please changes as fixups (can be intervened with commits)

martint commented on presto-main/src/main/java/io/prestosql/type/DateTimeOperators.java:
```diff
@@ -54,10 +54,7 @@ public static long datePlusIntervalDayToSecond(@SqlType(StandardTypes.DATE) long
-            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Cannot add hour, minutes or seconds to a date");
-        }
-        return TimeUnit.MILLISECONDS.toDays(interval) + date;
+        return datePlusIntervalDayToSecond(date, interval);
```
> We generally avoid UDF impl calling other UDF, to safeguard against incorrect error messages.

There's no hard and fast rule about that. I want to balance maintainability (i.e., having to fix bugs twice, change things twice, etc) with an explosion of methods (having 3 methods for everything seems overkill), especially when there's no validation involved. If there is, of course, it should be structured such that the message is appropriate for each implementation (e.g., see `dateMinusIntervalDayToSecond`). In this case, most methods don't do any validation.

> Also, datePlusIntervalDayToSecond and intervalDayToSecondPlusDate could produce different messages (when millis of day != 0), reversing param order.

It doesn't and it shouldn't. `DATE 'xxxx-xxx-xxx' + INTERVAL 'xxx'` should not produce a different message than `INTERVAL 'xxx' + DATE 'xxxx-xxx-xxx'`. In both cases, an interval is being added to a date. The order of the operands is irrelevant.

> I am leaning toward "validate inline and shared method for the math".

I agree, but only when it's deserved.

martint commented on presto-main/src/main/java/io/prestosql/sql/analyzer/ExpressionAnalyzer.java:
```diff
@@ -824,25 +824,21 @@ protected Type visitTimeLiteral(TimeLiteral node, StackableAstVisitorContext<Con
                 else {
-                    parseTimestampLiteral(node.getValue());
+                    type = TIMESTAMP;
+                    parseTimestamp(node.getValue());
```
Good catch. Too bad it wasn't caught by any tests

martint committed:
Add test for invalid legacy timestamp literal

martint committed:
Fix parsing of time in Kafka and Redis

parseTimeLiteral handles both time w/o timezone and time w/ timezone and
returns a value of the corresponding type. This is not appropriate for
this callsite, since we know the value has to be of type time w/o timezone.

martint committed:
Reorder methods in DateTimeUtils

findepi commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/util/KafkaLoader.java:
```diff
@@ -143,7 +143,7 @@ private Object convertValue(Object value, Type type)
             }
             if (TIME.equals(type)) {
-                return ISO8601_FORMATTER.print(parseTimeLiteral(timeZoneKey, (String) value));
+                return ISO8601_FORMATTER.print(parseTimeWithoutTimeZone(timeZoneKey, (String) value));
```
Same in RedisLoader (there is same code there)

findepi commented on presto-main/src/main/java/io/prestosql/util/DateTimeUtils.java:
```diff
@@ -196,35 +184,42 @@ public static long convertToTimestamp(String value)
-     * @deprecated applicable in legacy timestamp semantics only
+     * @return stack representation of TIMESTAMP type
      */
+    public static long parseTimestamp(String value)
```
I assume the changes in this class in `fixup: Clarify parsing of timestamps and time types` are just reorder.

martint replied:
Yes, it's easier to see once I squash the fixup commit.

findepi replied:
to be clear: I am not questioning this. just letting know i passed over this "without deep read"

martint replied:
Yup, got it. Take a look now. I squashed the fixups, so now it should be clear what the actual changes are in that original commit.

findepi commented on presto-main/src/test/java/io/prestosql/type/TestTimestampLegacy.java:
```diff
@@ -62,4 +69,17 @@ public void testCastFromSlice()
+    {
+        Session session = testSessionBuilder()
+                .setSystemProperty("legacy_timestamp", "true")
+                .setTimeZoneKey(TimeZoneKey.getTimeZoneKey("America/Bahia_Banderas"))
```
Use a zone different from test jvm zone (to cover the fact that jvm zone doesn't impact this).
eg Vilnius
https://github.com/prestosql/presto/blob/ca878d471740c07155ea29fb74fe8be0ba855a68/presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java#L129-L130

(i used Bahia in my previous comment, because i remember when some gap is in Bahia, sorry for misguiding you)

martint replied:
Does it matter? We rely on other tests to ensure that the session timezone is being respected. This is just making sure that "assuming session timezone is being respected => this timestamp is invalid". If the session timezone is not being respected and the test passes due to VM timezone, then that should trigger failures elsewhere.

findepi replied:
It does matter. For example timestamp could be detected as invalid if it's invalid in JVM zone, but then could be shifted if it is invalid in the session zone.

 Actually, i hope we also do have a test that parses timestamp being  a gap in JVM zone (but not in the session zone) and verifies it was succesful. We definitely do this in connector tests, but not sure if here.
